exploration/function-composition-part-1.md

```diff
@@ -0,0 +1,824 @@
+* In Model 2, some functions "look back" for the original value,
+(like `number`)
+while others return a new "source value"
+(like `getAge` in Example B1).
+
+The choice of internal value influences both areas,
+or rather, the desired answers to these questions
+constrain the choice of internal value.
+
+Composition might mean that "the second function operates on the output of the first",
+or "the second function operates on the **input** of the first, plus 'hints' supplied by the first",
+or it might mean either one depending on which function(s) are involved.
+
+The question is how to craft the spec in a way that is consistent with expectations.
+
+## Requirements
+
+_What properties does the solution have to manifest to enable the use-cases above?_
+
+In the rest of this document, we assume some version of Model 2.
+However, if Model 1 is more desired, the questions arise of how to
+forbid compositions of functions that would do surprising things
+under that model.
+
+Even under Model 2, some instances of composition won't make sense,
+so we need to define the error behavior when it doesn't make sense.
+
+This implies that we need:
+
+* Preservation of options
+* Preservation of the original operand value, at least for some functions
+* Passing through a new operand value _and_ options when appropriate, as with functions that extract fields
+(see Example B1)
+* (Possibly) Preservation of data about the names of functions that were previously called.
+
+The function registry needs to be explicit about which pieces of data
+a function preserves,
+and what it expects in its input.
+
+More generally than just "preservation of options",
+a solution needs to specify a minimum set of requirements
+for the internal value representation,
+so that functions can be passed the values they need.
+It also needs to provide a mechanism for declaring
+when functions can compose with each other.
+
+Other requirements:
+
+### Avoid over-constraining implementations
+
+An implementation is free to use any types
+and set of operations for coercing between types
+as long as it preserves the observable behavior
+of a message.
+
+The proposed solution should not impose
+unnecessary constraints on the implementation.
+However, it _should_ make the requirements
+for "resolved values"
+clear enough so that implementors can
+make a well-informed decision
+on what these types and operations should be.
+
+## Constraints
+
+_What prior decisions and existing conditions limit the possible design?_
+
+One prior decision is that the same definition of
+"resolved value" appears in multiple places in the spec.
+If "resolved value" is defined broadly enough
+(an annotated value with rich metadata),
+then this prior decision need not be changed.
+
+A second constraint is
+the difficulty of developing a precise definition of "resolved value"
+that can be made specific in the interface for custom functions,
+which is implementation-language-neutral.
```

Comment:
In general, I would be very happy for us to include a more concrete definition or description of "resolved values". Historically, that was one of the first part of the formatting spec that I tried to include, in #198. Unfortunately, that PR ended up being stuck for a year, before needing to be dismissed. The preceding lengthy discussion in #190 is also relevant background, though note of course that these discussions are from over two years ago, so will not fully reflect our current thinking.

Given this past history, I would find it useful to hear @mihnita's current thoughts on defining "resolved value" more explicitly, as he in particular was previously against the idea.

Reply:
I would also be interested to hear from @mihnita -- in terms of the implementations, I took inspiration from his code in defining a `FormattedPlaceholder` type that represents the return value of a function implementation, and trying to figure out how that type should actually work (in combination with #515) is how I got here.

Reply:
(and I'll take a look at those threads.)

Reply:
I tried to extract some of the top-level points from the lengthy discussion and added a "prior work" section: [fa625c3](https://github.com/unicode-org/message-format-wg/pull/753/commits/fa625c3185d34a3ad0a3505e37d01f0f945a5680).

If you think I've misrepresented anything or left out an important point, please let me know! I'm hoping that some of the questions that were challenging in those prior discussions might be easier to resolve now due to spec changes that have happened in the meantime. But not having been there for those discussions, I can't be certain of it.

Reply:
Re-reviewing this PR and the changes since my last review, I believe further progress on this is still blocked by waiting on input from @mihnita, as requested above. I would find it important to get buy-in on the general approach before spending more time on this.